src/Illuminate/Database/Eloquent/Model.php

```diff
@@ -1616,6 +1617,11 @@ public function newPivot(self $parent, array $attributes, $table, $exists, $usin
      */
     public function hasNamedScope($scope)
     {
+        if (method_exists($this, $scope)) {
+            $reflectedMethod = new \ReflectionMethod($this, $scope);
+            $returnType = $reflectedMethod->getReturnType();
+            return $returnType && $returnType->getName() === Scope::class;
```

Comment:
Could this be cached statically on the model?  It seems unnecessary to have to perform Reflection repeatedly if you call the scope multiple times within the life cycle of an app.

Also, if we're only checking the name of the return type, don't we lose the ability for userland to write their own child classes of Scope?

Reply:
Thanks for the feedback, I'll add the check for the child classes & look into caching the scope checks so that we only perform reflection once for scope methods. Will do it in the morning (it's late here).